src/GraphQL/Instrumentation/FieldMiddlewareBuilder.cs

```diff
@@ -7,63 +7,59 @@
 
 namespace GraphQL.Instrumentation
 {
-    public interface IFieldMiddlewareBuilder
-    {
-        IFieldMiddlewareBuilder Use(Func<FieldMiddlewareDelegate, FieldMiddlewareDelegate> middleware);
-
-        void ApplyTo(ISchema schema);
-    }
-
+    /// <summary>
+    /// Default implementation of <see cref="IFieldMiddlewareBuilder"/>.
+    /// </summary>
     public class FieldMiddlewareBuilder : IFieldMiddlewareBuilder
     {
-        private Func<FieldMiddlewareDelegate, FieldMiddlewareDelegate> _singleComponent;
-        private IList<Func<FieldMiddlewareDelegate, FieldMiddlewareDelegate>> _components;
+        private Func<ISchema, FieldMiddlewareDelegate, FieldMiddlewareDelegate> _singleMiddleware;
+        private IList<Func<ISchema, FieldMiddlewareDelegate, FieldMiddlewareDelegate>> _middlewares;
 
-        public IFieldMiddlewareBuilder Use(Func<FieldMiddlewareDelegate, FieldMiddlewareDelegate> middleware)
+        public IFieldMiddlewareBuilder Use(Func<ISchema, FieldMiddlewareDelegate, FieldMiddlewareDelegate> middleware)
         {
             middleware = middleware ?? throw new ArgumentNullException(nameof(middleware));
 
-            if (_singleComponent == null)
+            if (_singleMiddleware == null)
             {
                 // allocation free optimization for single middleware (InstrumentFieldsMiddleware)
-                _singleComponent = middleware;
+                _singleMiddleware = middleware;
             }
-            else if (_components == null)
+            else if (_middlewares == null)
             {
-                _components = new List<Func<FieldMiddlewareDelegate, FieldMiddlewareDelegate>>
+                _middlewares = new List<Func<ISchema, FieldMiddlewareDelegate, FieldMiddlewareDelegate>>
                 {
-                    _singleComponent,
+                    _singleMiddleware,
                     middleware
                 };
             }
             else
             {
-                _components.Add(middleware);
+                _middlewares.Add(middleware);
             }
 
             return this;
         }
 
-        public FieldMiddlewareDelegate Build(FieldMiddlewareDelegate start = null)
+        public FieldMiddlewareDelegate Build(FieldMiddlewareDelegate start = null, ISchema schema = null)
```

Comment:
Just curious, why is this function public at all?  Certainly if it is private, schema can be required, as it will always receive one from `ApplyTo`.

Reply:
I am sure that in this project there are a huge number of APIs that can and should be made private. I even saw a corresponding issue for this. This simplifies understanding of important APIs, and not important ones go into implementation details. Each case must be considered separately. I am not familiar with the history of the project so much as to understand 100% which APIs can be removed and which not. @joemcbride can we make this method `internal`?